Sprint 8 solution time and duration

ToNyQwW opened the pull request:
Привет)) произошла масштабная переработка кода. Сначала хотел просто добавлять задачи в приоритетный список, потом показалось прикольной идея вынести в отдельный класс. После я подумал, что те задачи, которые не могут попасть в приоритет, тоже бы хранить и если освободится место то добавлять в приоритетные) В итоге сидел пару дней и пытался это все реализовать)) + сделал дополнительные задания на поиск пересечений + optional. На тестах ловил кучу багов и всякой жести, но вроде все поправил, что смог найти))

ToNyQwW committed:
Изменение класса Task
добавлены поля startTime и duration
изменены конструкторы с учетом новых полей
добавлены геттеры и сеттеры на новые поля
добавлен метод getEndTime()
изменен toString()

ToNyQwW committed:
В классе Task:
Добавлена константа DEFAULT_ID
Изменил конструкторы;
Добавил методы startTimeToString() и durationToLong();
Внес изменения в toString().
В классе subTask:
Изменил конструкторы;
Изменил toString().

ToNyQwW committed:
Изменения в классе EpicTask:
добавил поле endTime;
переопределил метод getEndTime();
Внес изменения в конструкторы.

ToNyQwW committed:
В классе Main:
провел рефакторинг кода с учетом предыдущих коммитов;
В классе FileBackedTaskManager:
изменил метод fromString c учетом ранних изменений в subtask;
Внес изменения в тестах для корректной работы

ToNyQwW committed:
Создал новый класс ConverterToTask и перенес туда метод fromString() из класса FileBackedTaskManager;
Произвел рефакторинг в FileBackedTaskManagerTest

ToNyQwW committed:
В классе FileBackedTaskManager:
Создал метод getSortedTasks();
изменил методы loadFromFile и save.
Создал новое исключение ConvertToTaskException;
Переработал метод fromString() в классе ConverterToTask;

ToNyQwW committed:
В классе EpicTask добавил метод setEndTime().
В классе ConverterToTask произвел рефакторинг метода fromString().
В классе InMemoryTaskManager добавил метод updateEpicTaskTime().

ToNyQwW committed:
В классе EpicTask добавил метод setEndTime().
В классе ConverterToTask произвел рефакторинг метода fromString().
В классе InMemoryTaskManager добавил методы updateEpicTaskTime() и setNullTimeInEpicTask().

ToNyQwW committed:
Добавлены тесты на метод updateEpicTaskTime

ToNyQwW committed:
Добавлен тесты на ConverterToTask

ToNyQwW committed:
В классе InMemoryTaskManager:
добавлено поле prioritizedTasks;
добавлены методы для работы с новым полем;
изменены другие методы класса под работу с новым функционалом.

ToNyQwW committed:
В классе Task добавлен метод isTaskWithTime().
В Классе InMemoryTaskManager добавлен метод removeFromPrioritizedTasks() и переработана работа класса с учетом нового метода

ToNyQwW committed:
Вынес set prioritizedTasks и всю логику работы с ним в отдельный класс PriorityTaskManager

ToNyQwW committed:
Merge remote-tracking branch 'origin/sprint_8-solution-time-and-duration' into sprint_8-solution-time-and-duration

# Conflicts:
#	src/service/InMemoryTaskManager.java

ToNyQwW committed:
Revert "Вынес set prioritizedTasks и всю логику работы с ним в отдельный класс PriorityTaskManager"

This reverts commit 065205cb436ac68f8a09f4656564d1e32ff5446c.

ToNyQwW committed:
Revert "Вынес set prioritizedTasks и всю логику работы с ним в отдельный класс PriorityTaskManager"

This reverts commit 065205cb

ToNyQwW committed:
Merge remote-tracking branch 'origin/sprint_8-solution-time-and-duration' into sprint_8-solution-time-and-duration

ToNyQwW committed:
Чиню коммиты))

ToNyQwW committed:
Чиню коммиты))

ToNyQwW committed:
Изменил возвращаемое значение методов поиска задач

ToNyQwW committed:
Добавлен новый класс TaskTimeIntersectionChecker для реализации доп задания;
переработан PriorityTaskManager под новый класс

ToNyQwW committed:
Merge remote-tracking branch 'origin/sprint_8-solution-time-and-duration' into sprint_8-solution-time-and-duration

# Conflicts:
#	src/service/InMemoryTaskManager.java

ToNyQwW committed:
Фиксы, доп тесты %_%

GeorgeGolikov commented on src/service/PriorityTaskManager.java:
```diff
@@ -0,0 +1,62 @@
+import java.time.LocalDateTime;
+import java.util.*;
+
+public class PriorityTaskManager implements PriorityManager {
```
отличная идея с отдельным менеджером)

GeorgeGolikov commented on src/service/FileBackedTaskManager.java:
```diff
@@ -1,80 +1,76 @@
         }
     }
 
+    private TreeSet<Task> getSortedTasks() {
```
у тебя ведь уже есть отсортированный список задач в PriorityManager - можно воспользоваться им, чтобы не пересчитывать список заново при сохранении

GeorgeGolikov reviewed (commented):
Привет! Супер, вот это хард уровень работа) Те задачи которые не попали необязательно было конечно делать, но ты красавчик! Главное, что получилось всё отладить и довести до конца - это важно в сложных работах.

Оставил небольшое замечание

ToNyQwW committed:
Добавил getUnprioritizedTasks в классе InMemoryTaskManager;
Фикс бага в методах удаления задач/подзадач (не удалялись задачи из unPriority);
добавлены тесты на удаление задачи из unPriority

GeorgeGolikov reviewed (approved):
Без меня заметил ошибку)) Круто! Прям большой молодец
